finality-aleph/src/sync/forest/mod.rs

```diff
@@ -289,51 +305,102 @@ impl<I: PeerId, J: Justification> Forest<I, J> {
             self.prune(&id);
         }
         self.compost_bin.retain(|k| k.number() > level);
+        self.justified_blocks.retain(|k, _| k > &level);
     }
 
     /// Attempt to finalize one block, returns the correct justification if successful.
-    pub fn try_finalize(&mut self) -> Option<J> {
-        for child_id in self.root_children.clone().into_iter() {
-            if let Some(VertexWithChildren { vertex, children }) = self.vertices.remove(&child_id) {
+    pub fn try_finalize(&mut self, number: &u32) -> Option<J> {
+        // cached as ready
+        if let Some(id) = self.justified_blocks.get(number) {
+            if let Some(VertexWithChildren { vertex, children }) = self.vertices.remove(id) {
                 match vertex.ready() {
+                    // ready indeed
                     Ok(justification) => {
-                        self.root_id = child_id;
+                        self.root_id = id.clone();
                         self.root_children = children;
                         self.prune_level(self.root_id.number());
                         return Some(justification);
                     }
+                    // cache corrupted, we'll wait for it to fix itself
```

Comment:
you want to keep justified vertices in a separate hashmap? but this still leads to a similar situation?

Reply:
Not really, since the code in this branch will no longer be dead – if you have a justified vertex, that's nonetheless not finalizable (block not imported) then this branch will actually be executed.

Reply:
moving justifications out of `Vertex` requires another large refactor, and rewriting the state logic - do we really need to do this?
keeping justifications in `Vertex` and separating justified/unjustified vertices into two hashmaps will lead to a similar situation, where we need to check if a `Vertex` in the justified collection indeed has a justification, and there'll be some dead code to handle the incorrect case

Reply:
right?